### PR TITLE
Dynamics criterion (AcceleratorTools part 1)

### PR DESCRIPTION
**Pull request by mjohnson541**
This pull request implements the dynamics criterion.  The dynamic criterion is designed to work alongside the flux criterion, but instead considers how edge reactions affect core concentrations rather than the formation rates of edge species.  The dynamics criterion moves edge reactions (all species in a given edge reaction) rather than a single edge species to the core.

### CODE REVIEW

File: rmgpy/solver/base.pyx [hunk 1]
```diff
@@ -493,7 +504,70 @@ cdef class ReactionSystem(DASx):
             for index in xrange(numPdepNetworks):
                 if maxNetworkLeakRateRatios[index] < networkLeakRateRatios[index]:
                     maxNetworkLeakRateRatios[index] = networkLeakRateRatios[index]
-
+            
+            #get abs(delta(Ln(total accumulation numbers))) (accumulation number=Production/Consumption)
+            #(the natural log operation is avoided until after the maximum accumulation number is found)
+            totalDivAccumNums = numpy.ones(numEdgeReactions)
```
Comment (mliu49) [inline rmgpy/solver/base.pyx]: What is `totalDivAccumNums` short for?

File: rmgpy/solver/base.pyx [hunk 2]
```diff
@@ -493,7 +504,70 @@ cdef class ReactionSystem(DASx):
...
+                    for index in xrange(numEdgeReactions):
+                        maxEdgeReactionAccum = 0.0
+                        maxAccumReactionIndex = -1
+                        if infAccumNumIndex in self.productIndices[index+numCoreReactions]:
+                            reactionRate = edgeReactionRates[index]
+                            if reactionRate > maxEdgeReactionAccum:
+                                maxEdgeReactionAccum = reactionRate
+                                maxAccumReactionIndex = index
+                    maxAccumReaction = edgeReactions[maxAccumReactionIndex]
+                elif len(totalDivAccumNums)>0:
```
Comment (mliu49) [inline rmgpy/solver/base.pyx]: Could this just be `else:`?

File: rmgpy/solver/base.pyx [hunk 3]
```diff
@@ -610,11 +696,12 @@ cdef class ReactionSystem(DASx):
         # (if the simulation was valid)
         return terminated, invalidObject
 
-    cpdef logRates(self, double charRate, object species, double speciesRate, object network, double networkRate):
+    cpdef logRates(self, double charRate, object species, double speciesRate, double maxDifLnAccumNum, object network, double networkRate):
         """
         Log information about the current maximum species and network rates.
         """
         logging.info('    Characteristic rate: {0:10.4e} mol/m^3*s'.format(charRate))
+        logging.info('    Max(Dif(Ln(accumulation number))):  {0:10.4e}'.format(maxDifLnAccumNum))
```
Comment (mliu49) [inline rmgpy/solver/base.pyx]: This label isn't particularly meaningful to me. Is there a different description you could use?

Comment (mjohnson541) [inline rmgpy/solver/base.pyx]: This has been changed to "Dynamics number"

File: rmgpy/solver/base.pyx (hunk 2 above)
Comment (mjohnson541) [inline rmgpy/solver/base.pyx]: fixed

File: rmgpy/solver/base.pyx (hunk 1 above)
Comment (mjohnson541) [inline rmgpy/solver/base.pyx]: product of the ratios of the accumulation numbers (for a given reaction)

File: rmgpy/solver/base.pyx [hunk 4]
```diff
@@ -363,23 +366,27 @@ cdef class ReactionSystem(DASx):
...
         cdef int numCoreSpecies, numEdgeSpecies, numPdepNetworks, numCoreReactions
-        cdef double stepTime, charRate, maxSpeciesRate, maxNetworkRate
+        cdef double stepTime, charRate, maxSpeciesRate, maxNetworkRate, maxEdgeReactionAccum
         cdef numpy.ndarray[numpy.float64_t, ndim=1] y0 #: Vector containing the number of moles of each species
-        cdef numpy.ndarray[numpy.float64_t, ndim=1] coreSpeciesRates, edgeSpeciesRates, networkLeakRates
+        cdef numpy.ndarray[numpy.float64_t, ndim=1] coreSpeciesRates, edgeSpeciesRates, networkLeakRates, coreSpeciesProductionRates, coreSpeciesConsumptionRates, totalDivAccumNums
         cdef numpy.ndarray[numpy.float64_t, ndim=1] maxCoreSpeciesRates, maxEdgeSpeciesRates, maxNetworkLeakRates,maxEdgeSpeciesRateRatios, maxNetworkLeakRateRatios
         cdef bint terminated
-        cdef object maxSpecies, maxNetwork
+        cdef object maxSpecies, maxNetwork,
```
Comment (mjohnson541) [inline rmgpy/solver/base.pyx]: done

File: rmgpy/solver/base.pyx [hunk 5]
```diff
@@ -346,8 +349,8 @@ cdef class ReactionSystem(DASx):
 
     @cython.boundscheck(False)
     cpdef simulate(self, list coreSpecies, list coreReactions, list edgeSpecies, list edgeReactions,
-        double toleranceKeepInEdge, double toleranceMoveToCore, double toleranceInterruptSimulation,
-        list pdepNetworks=None, absoluteTolerance=1e-16, relativeTolerance=1e-8, sensitivity=False, 
+        double toleranceKeepInEdge, double toleranceMoveToCore, double toleranceInterruptSimulation,double toleranceReactionMoveToCore=numpy.inf, double toleranceReactionInterruptSimulation=numpy.inf,
```
Comment (mjohnson541) [inline rmgpy/solver/base.pyx]: done

File: rmgpy/solver/base.pxd [hunk 6]
```diff
@@ -106,10 +107,11 @@ cdef class ReactionSystem(DASx):
         list pdepNetworks=?, atol=?, rtol=?, sensitivity=?, sens_atol=?, sens_rtol=?, filterReactions=?)
 
     cpdef simulate(self, list coreSpecies, list coreReactions, list edgeSpecies, list edgeReactions,
-        double toleranceKeepInEdge, double toleranceMoveToCore, double toleranceInterruptSimulation,
-        list pdepNetworks=?, absoluteTolerance=?, relativeTolerance=?, sensitivity=?, sensitivityAbsoluteTolerance=?, 
-        sensitivityRelativeTolerance=?, sensWorksheet=?, filterReactions=?)
-
-    cpdef logRates(self, double charRate, object species, double speciesRate, object network, double networkRate)
+        double toleranceKeepInEdge, double toleranceMoveToCore,  double toleranceInterruptSimulation,double toleranceReactionMoveToCore=?,double toleranceReactionInterruptSimulation=?,
```
Comment (mjohnson541) [inline rmgpy/solver/base.pxd]: done

File: rmgpy/rmg/main.py [hunk 7]
```diff
@@ -608,7 +616,18 @@ def execute(self, **kwargs):
...
                         # Store the maximum leak species along with the associated network
                         obj = (obj, obj.getMaximumLeakSpecies(reactionSystem.T.value_si, reactionSystem.P.value_si))
-                    objectsToEnlarge.append(obj)
+                        objectsToEnlarge.append(obj)
+                    elif isinstance(obj, Species):
+                        objectsToEnlarge.append(obj)
+                        assert len(objectsToEnlarge)>0
+                    elif isinstance(obj,Reaction):
+                        potentialSpcs = obj.reactants+obj.products
+                        filterFcn = lambda x: not ((x in self.reactionModel.core.species)) #remove species already in core
```
Comment (mjohnson541) [inline rmgpy/rmg/main.py]: done

**Pull request merged**